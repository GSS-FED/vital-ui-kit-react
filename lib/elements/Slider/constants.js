export default {
  small: {
    handlerSize: '24px',
  },
  medium: {
    handlerSize: '26px',
  },
  large: {
    handlerSize: '28px',
  }
};
