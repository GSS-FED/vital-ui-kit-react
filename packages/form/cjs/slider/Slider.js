'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var React = _interopRequireWildcard(_react);

var _resizeObserverPolyfill = require('resize-observer-polyfill');

var _resizeObserverPolyfill2 = _interopRequireDefault(_resizeObserverPolyfill);

var _reactIcon = require('@vital-ui/react-icon');

var _reactIcon2 = _interopRequireDefault(_reactIcon);

var _reactTooltip = require('@vital-ui/react-tooltip');

var _reactTooltip2 = _interopRequireDefault(_reactTooltip);

var _styled = require('./styled');

var _track = require('./track');

var _track2 = _interopRequireDefault(_track);

var _constants = require('./constants');

var _constants2 = _interopRequireDefault(_constants);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * 
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Copyright Â© 2018 Galaxy Software Services https://github.com/GSS-FED/vital-ui-kit-react
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * MIT license
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */

/**
 * @render react
 * @name Slider
 * @description Define the value by dragging the handle or side buttons
 * @example
 * <Slider
 *  size="large"
 *  hasButton
 *  value={50}
 *  max={100}
 *  min={0}
 *  step={5}
 * />
 */
var Slider = function (_React$Component) {
  _inherits(Slider, _React$Component);

  function Slider() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Slider);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Slider.__proto__ || Object.getPrototypeOf(Slider)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      active: false,
      position: -_constants2.default[_this.props.size].handlerSize / 2,
      // limit of the handle drag
      limit: 0,
      // handle offset
      grab: 0,
      // cache the mouse down postion x
      startX: 0,
      value: _this.props.value || 0
    }, _this.start = 700, _this.timeout = undefined, _this.handleUpdate = function () {
      if (!_this.slider || !_this.track || !_this.handle) {
        return;
      }
      var trackWidth = _this.track.offsetWidth;
      var handleWith = _this.handle.offsetWidth;
      _this.setState(function (prevState) {
        return {
          limit: trackWidth - handleWith,
          grab: -handleWith / 2,
          position: _this.getPositionFromValue(prevState.value)
        };
      });
    }, _this.handleStart = function (e) {
      if (_this.props.disabled) return;
      if (_this.props.onChangeStart) _this.props.onChangeStart(e);
      _this.setState({
        active: true,
        startX: e.clientX
      });
      // $FlowFixMe
      document.addEventListener('mousemove', _this.handleDrag);
      // $FlowFixMe
      document.addEventListener('mouseup', _this.handleEnd);
    }, _this.handleTrack = function (e) {
      if (_this.props.disabled) return;
      e.stopPropagation();
      // $FlowFixMe
      var percentage = e.nativeEvent.offsetX / _this.track.clientWidth;
      var value = _this.props.step * Math.round(percentage * (_this.props.max - _this.props.min) / _this.props.step);
      _this.setState({ value: value });
      if (_this.props.onChange) _this.props.onChange(value, e);
    }, _this.handleDrag = function (e) {
      e.stopPropagation();
      var value = _this.getValueFromPosition(e.clientX);
      _this.setState({ value: value });
      if (_this.props.onChange) _this.props.onChange(value, e);
    }, _this.handleEnd = function (e) {
      if (_this.props.onChangeComplete) _this.props.onChangeComplete(e);
      _this.setState(function (prevState) {
        return {
          active: false,
          position: _this.getPositionFromValue(prevState.value)
        };
      });
      // $FlowFixMe
      document.removeEventListener('mousemove', _this.handleDrag);
      // $FlowFixMe
      document.removeEventListener('mouseup', _this.handleEnd);
    }, _this.handleIncrease = function () {
      if (_this.props.disabled) return;
      _this.increaseByStep();
      _this.timeout = setTimeout(_this.handleIncrease, _this.start);
      _this.start = _this.start / 2;
    }, _this.handleDecrease = function () {
      if (_this.props.disabled) return;
      _this.decreaseByStep();
      _this.timeout = setTimeout(_this.handleDecrease, _this.start);
      _this.start = _this.start / 2;
    }, _this.increaseByStep = function () {
      _this.setState(function (prevState) {
        return {
          value: _this.clamp(prevState.value + _this.props.step, _this.props.max, _this.props.min)
        };
      });
    }, _this.decreaseByStep = function () {
      _this.setState(function (prevState) {
        return {
          value: _this.clamp(prevState.value - _this.props.step, _this.props.max, _this.props.min)
        };
      });
    }, _this.getPositionFromValue = function (value) {
      var _this$props = _this.props,
          min = _this$props.min,
          max = _this$props.max;
      var limit = _this.state.limit;

      var diffMaxMin = max - min;
      var diffValMin = value - min;
      var percentage = diffValMin / diffMaxMin;
      var pos = Math.round(percentage * limit);
      return pos;
    }, _this.getValueFromPosition = function (pos) {
      var _this$state = _this.state,
          startX = _this$state.startX,
          limit = _this$state.limit,
          grab = _this$state.grab,
          position = _this$state.position;

      var diff = pos - startX;

      var newHandlerLeft = _this.clamp(position + diff, limit, grab);
      var percentage = (newHandlerLeft - grab) / (limit - grab);
      var value = _this.props.step * Math.round(percentage * (_this.props.max - _this.props.min) / _this.props.step);
      return value;
    }, _this.clamp = function (value, max, min) {
      return Math.min(Math.max(value, min), max);
    }, _this.renderDecreaseButton = function () {
      var buttonShow = _this.props.hasButton;
      if (_this.props.decreaseButton) buttonShow = true;
      if (buttonShow) {
        var buttonProps = {
          onMouseDown: _this.handleDecrease,
          onMouseUp: function onMouseUp() {
            clearTimeout(_this.timeout);
            _this.start = 700;
          },
          style: { marginRight: '12px', flex: '0 0 auto' }
        };
        if (_this.props.decreaseButton) {
          return React.cloneElement(_this.props.decreaseButton, _extends({}, buttonProps));
        }
        return React.createElement(
          _styled.Button,
          _extends({}, buttonProps, { circle: true, size: 'xsmall' }),
          React.createElement(_reactIcon2.default, { name: 'chevron-left', size: '13' })
        );
      }
      return null;
    }, _this.renderIncreaseButton = function () {
      var buttonShow = _this.props.hasButton;
      if (_this.props.increaseButton) buttonShow = true;
      if (buttonShow) {
        var buttonProps = {
          onMouseDown: _this.handleIncrease,
          onMouseUp: function onMouseUp() {
            clearTimeout(_this.timeout);
            _this.start = 700;
          },
          style: { marginLeft: '12px', flex: '0 0 auto' }
        };
        if (_this.props.increaseButton) {
          return React.cloneElement(_this.props.increaseButton, _extends({}, buttonProps));
        }
        return React.createElement(
          _styled.Button,
          _extends({}, buttonProps, { circle: true, size: 'xsmall' }),
          React.createElement(_reactIcon2.default, { name: 'chevron-right', size: '13' })
        );
      }
      return null;
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  // delay timeout of the button calling function


  _createClass(Slider, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.handleUpdate();
      var resizeObserver = new _resizeObserverPolyfill2.default(this.handleUpdate);
      resizeObserver.observe(this.track);
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      //
      var position = this.getPositionFromValue(this.state.value);

      return React.createElement(
        _styled.Root,
        {
          innerRef: function innerRef(s) {
            _this2.slider = s;
          },
          disabled: this.props.disabled,
          'aria-valuemin': this.props.min,
          'aria-valuemax': this.props.max,
          'aria-valuenow': this.props.value
        },
        this.renderDecreaseButton(),
        React.createElement(
          _styled.Wrapper,
          {
            size: this.props.size,
            disabled: this.props.disabled
          },
          React.createElement(_track2.default, {
            size: this.props.size,
            onMouseDown: this.handleTrack,
            onMouseUp: function onMouseUp(e) {
              _this2.setState({ startX: e.clientX });
              _this2.handleEnd(e);
            },
            onTouchStart: this.handleStart,
            onTouchEnd: this.handleEnd,
            trackRef: function trackRef(s) {
              _this2.track = s;
            },
            disabled: this.props.disabled,
            selectionWidth: position - this.state.grab
          }),
          React.createElement(
            _reactTooltip2.default,
            {
              placement: 'bottom',
              overlay: this.state.value,
              trigger: ['hover']
            },
            React.createElement(_styled.Handler, {
              size: this.props.size,
              innerRef: function innerRef(s) {
                _this2.handle = s;
              },
              style: {
                left: position + 'px'
              },
              active: this.state.active,
              onMouseDown: this.handleStart
            })
          )
        ),
        this.renderIncreaseButton()
      );
    }
  }]);

  return Slider;
}(React.Component);

Slider.defaultProps = {
  size: 'medium',
  disabled: false,
  hasButton: false,
  decreaseButton: null,
  increaseButton: null,
  trackLabel: false,
  onChangeStart: null,
  onChangeComplete: null
};
exports.default = Slider;