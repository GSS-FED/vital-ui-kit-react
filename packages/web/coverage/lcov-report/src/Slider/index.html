<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/Slider</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> src/Slider
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">43.27% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>45/104</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">40.54% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>15/37</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">50% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>17/34</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">47.78% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>43/90</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line low'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file low" data-value="Slider.jsx"><a href="Slider.jsx.html">Slider.jsx</a></td>
	<td data-value="35.87" class="pic low"><div class="chart"><div class="cover-fill" style="width: 35%;"></div><div class="cover-empty" style="width:65%;"></div></div></td>
	<td data-value="35.87" class="pct low">35.87%</td>
	<td data-value="92" class="abs low">33/92</td>
	<td data-value="40" class="pct low">40%</td>
	<td data-value="35" class="abs low">14/35</td>
	<td data-value="34.62" class="pct low">34.62%</td>
	<td data-value="26" class="abs low">9/26</td>
	<td data-value="39.74" class="pct low">39.74%</td>
	<td data-value="78" class="abs low">31/78</td>
	</tr>

<tr>
	<td class="file empty" data-value="constants.js"><a href="constants.js.html">constants.js</a></td>
	<td data-value="0" class="pic empty"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	</tr>

<tr>
	<td class="file empty" data-value="index.js"><a href="index.js.html">index.js</a></td>
	<td data-value="0" class="pic empty"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	</tr>

<tr>
	<td class="file high" data-value="styled.js"><a href="styled.js.html">styled.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="12" class="abs high">12/12</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="2" class="abs medium">1/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="12" class="abs high">12/12</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Tue Jun 05 2018 19:42:58 GMT+0800 (CST)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
<script src="../../block-navigation.js"></script>
</body>
</html>
