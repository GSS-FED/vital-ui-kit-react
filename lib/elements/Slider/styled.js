import styled from 'styled-components';
import ButtonBase from '../Button';
import constants from './constants';

const Root = styled.div`
  display: flex;
  position: relative;
  min-width: 200px;
  align-items: center;

  ${props =>
    props.disabled &&
    `
    opacity: .6;
  `};
`;

const Wrapper = styled.div`
  display: block;
  position: relative;
  width: 100%;
  height: ${props => constants[props.size].handlerSize};
`;

const Track = styled.div`
  position: absolute;
  margin: 0;
  padding: 0;
  cursor: pointer;
  top: 50%;
  left: 0;
  height: ${props => constants[props.size].trackHeight};
  margin-top: -4px;
  background-repeat: repeat-x;
  border-radius: 4px;
  border-color: #c4d2eb;
  width: 100%;
  background-color: #d8e3f6;
  cursor: pointer;
`;

const Selection = styled.div`
  position: absolute;
  margin: 0;
  padding: 0;
  border-radius: 4px;
  background-color: #0e86fe;
  left: 0;
  height: ${props => constants[props.size].trackHeight};
  margin-top: -4px;
  top: 50%;
  width: ${props => `${props.width}px`};
  pointer-events: none;
`;

const Handler = styled.div`
  left: ${props => `${props.left}px`}
  position: absolute;
  border-radius: 100%;
  border-color: #c4d2eb;
  background-color: #ffffff;
  box-shadow: 0 0 0 1px rgba(196, 210, 235, 0.3);
  text-indent: -3333px;
  overflow: hidden;
  text-decoration: none;
  text-align: center;
  outline: 0;
  border-style: solid;
  border-width: 1px;
  width: ${props =>
     constants[props.size].handlerSize};
  height: ${props => constants[props.size].handlerSize};
  cursor: pointer;

  &:hover {
    border-color: #C4D2EB;
    background-color: #FFFFFF;
    box-shadow: 0 0 0 2px rgba(196, 210, 235, 0.3);
  }
`;

const Button = styled(ButtonBase)`
  flex: 0 0 auto;
  width: 24px;
  height: 24px;
`;

export {
  Root,
  Wrapper,
  Track,
  Selection,
  Handler,
  Button,
}
