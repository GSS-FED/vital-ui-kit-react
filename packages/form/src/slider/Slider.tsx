/**
 * Copyright Â© 2018 Galaxy Software Services https://github.com/GSS-FED/vital-ui-kit-react
 * MIT license
 */

import * as React from 'react';
import ResizeObserver from 'resize-observer-polyfill';
import {} from '@vital-ui/react-icon';
import { Tooltip } from '@vital-ui/react-tooltip';

import {
  SliderRoot,
  SliderWrapper,
  SlideHandler,
  Button,
} from './styled';
import Track from './track';

import constants from './constants';

type State = {
  active: boolean;
  position: number;
  grab: number;
  limit: number;
  startX: number;
  value: number;
};

export interface SliderProps {
  size?: 'large' | 'medium' | 'small';
  min: number;
  max: number;
  step: number;
  value: number;
  disabled?: boolean;
  hasButton?: boolean;
  decreaseButton?: React.ReactElement<any>;
  increaseButton?: React.ReactElement<any>;
  trackLabel?: boolean;
  onChange?: (
    value: number,
    e: React.SyntheticEvent<HTMLElement>,
  ) => void;
  onChangeStart?: React.EventHandler<any>;
  onChangeComplete?: React.EventHandler<any>;
}

/**
 * @render react
 * @name Slider
 * @description Define the value by dragging the handle or side buttons
 * @example
 * <Slider
 *  size="large"
 *  hasButton
 *  value={50}
 *  max={100}
 *  min={0}
 *  step={5}
 * />
 */
export class Slider extends React.Component<SliderProps, State> {
  static defaultProps = {
    size: 'medium',
    disabled: false,
    hasButton: false,
    decreaseButton: null,
    increaseButton: null,
    trackLabel: false,
    onChangeStart: null,
    onChangeComplete: null,
  };

  state = {
    active: false,
    position: -constants[this.props.size!].handlerSize / 2,
    // limit of the handle drag
    limit: 0,
    // handle offset
    grab: 0,
    // cache the mouse down postion x
    startX: 0,
    value: this.props.value || 0,
  };

  slider = React.createRef<HTMLDivElement>();

  track = React.createRef<HTMLDivElement>();

  handle = React.createRef<HTMLDivElement>();

  // delay timeout of the button calling function
  start = 700;

  timeout: any = undefined;

  componentDidMount() {
    this.handleUpdate();
    const resizeObserver = new ResizeObserver(this.handleUpdate);
    if (this.track.current) {
      resizeObserver.observe(this.track.current);
    }
  }

  handleUpdate = () => {
    if (
      !this.slider.current ||
      !this.track.current ||
      !this.handle.current
    ) {
      return;
    }
    const trackWidth = this.track.current.offsetWidth;
    const handleWith = this.handle.current.offsetWidth;
    this.setState(prevState => ({
      limit: trackWidth - handleWith,
      grab: -handleWith / 2,
      position: this.getPositionFromValue(prevState.value),
    }));
  };

  handleStart: React.ReactEventHandler = e => {
    if (this.props.disabled) {
      return;
    }
    if (this.props.onChangeStart) {
      this.props.onChangeStart(e);
    }
    this.setState({
      active: true,
      // @ts-ignore
      startX: e.clientX,
    });
    // @ts-ignore
    document.addEventListener('mousemove', this.handleDrag);
    // @ts-ignore
    document.addEventListener('mouseup', this.handleEnd);
  };

  handleTrack: React.MouseEventHandler<HTMLElement> = e => {
    if (this.props.disabled || !this.track.current) {
      return;
    }
    e.stopPropagation();
    const percentage =
      e.nativeEvent.offsetX / this.track.current.clientWidth;
    const value =
      this.props.step *
      Math.round(
        (percentage * (this.props.max - this.props.min)) /
          this.props.step,
      );
    this.setState({ value });
    if (this.props.onChange) {
      this.props.onChange(value, e);
    }
  };

  handleDrag: React.MouseEventHandler<HTMLElement> = e => {
    e.stopPropagation();
    const value = this.getValueFromPosition(e.clientX);
    this.setState({ value });
    if (this.props.onChange) {
      this.props.onChange(value, e);
    }
  };

  handleEnd: React.ReactEventHandler = e => {
    if (this.props.onChangeComplete) {
      this.props.onChangeComplete(e);
    }
    this.setState(prevState => ({
      active: false,
      position: this.getPositionFromValue(prevState.value),
    }));
    // @ts-ignore
    document.removeEventListener('mousemove', this.handleDrag);
    // @ts-ignore
    document.removeEventListener('mouseup', this.handleEnd);
  };

  handleIncrease = () => {
    if (this.props.disabled) {
      return;
    }
    this.increaseByStep();
    this.timeout = setTimeout(this.handleIncrease, this.start);
    this.start = this.start / 2;
  };

  handleDecrease = () => {
    if (this.props.disabled) {
      return;
    }
    this.decreaseByStep();
    this.timeout = setTimeout(this.handleDecrease, this.start);
    this.start = this.start / 2;
  };

  increaseByStep = () => {
    this.setState(prevState => ({
      value: this.clamp(
        prevState.value + this.props.step,
        this.props.max,
        this.props.min,
      ),
    }));
  };

  decreaseByStep = () => {
    this.setState(prevState => ({
      value: this.clamp(
        prevState.value - this.props.step,
        this.props.max,
        this.props.min,
      ),
    }));
  };

  getPositionFromValue = (value: number): number => {
    const { min, max } = this.props;
    const { limit } = this.state;
    const diffMaxMin = max - min;
    const diffValMin = value - min;
    const percentage = diffValMin / diffMaxMin;
    const pos = Math.round(percentage * limit);
    return pos;
  };

  getValueFromPosition = (pos: number): number => {
    const { startX, limit, grab, position } = this.state;
    const diff = pos - startX;

    const newHandlerLeft = this.clamp(position + diff, limit, grab);
    const percentage = (newHandlerLeft - grab) / (limit - grab);
    const value =
      this.props.step *
      Math.round(
        (percentage * (this.props.max - this.props.min)) /
          this.props.step,
      );
    return value;
  };

  clamp = (value: number, max: number, min: number) =>
    Math.min(Math.max(value, min), max);

  renderDecreaseButton = () => {
    let buttonShow = this.props.hasButton;
    if (this.props.decreaseButton) {
      buttonShow = true;
    }
    if (buttonShow) {
      const buttonProps = {
        onMouseDown: this.handleDecrease,
        onMouseUp: () => {
          clearTimeout(this.timeout);
          this.start = 700;
        },
        style: { marginRight: '12px', flex: '0 0 auto' },
      };
      if (this.props.decreaseButton) {
        return React.cloneElement(this.props.decreaseButton, {
          ...buttonProps,
        });
      }
      return (
        <Button {...buttonProps} circle size="xsmall">
          <svg
            xmlns="http://www.w3.org/2000/svg"
            aria-hidden="true"
            data-prefix="fas"
            data-icon="chevron-left"
            role="img"
            viewBox="0 0 320 512"
          >
            <path
              fill="currentColor"
              d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
            />
          </svg>
        </Button>
      );
    }
    return null;
  };

  renderIncreaseButton = () => {
    let buttonShow = this.props.hasButton;
    if (this.props.increaseButton) {
      buttonShow = true;
    }
    if (buttonShow) {
      const buttonProps = {
        onMouseDown: this.handleIncrease,
        onMouseUp: () => {
          clearTimeout(this.timeout);
          this.start = 700;
        },
        style: { marginLeft: '12px', flex: '0 0 auto' },
      };
      if (this.props.increaseButton) {
        return React.cloneElement(this.props.increaseButton, {
          ...buttonProps,
        });
      }
      return (
        <Button {...buttonProps} circle size="xsmall">
          <svg
            xmlns="http://www.w3.org/2000/svg"
            aria-hidden="true"
            data-prefix="fas"
            data-icon="chevron-right"
            role="img"
            viewBox="0 0 320 512"
          >
            <path
              fill="currentColor"
              d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
            />
          </svg>
        </Button>
      );
    }
    return null;
  };

  render() {
    //
    const position = this.getPositionFromValue(this.state.value);

    return (
      <SliderRoot
        innerRef={this.slider}
        disabled={this.props.disabled!}
        aria-valuemin={this.props.min}
        aria-valuemax={this.props.max}
        aria-valuenow={this.props.value}
      >
        {this.renderDecreaseButton()}
        <SliderWrapper
          size={this.props.size!}
          disabled={this.props.disabled!}
        >
          <Track
            size={this.props.size}
            onMouseDown={this.handleTrack}
            onMouseUp={(e: React.MouseEvent<any>) => {
              this.setState({ startX: e.clientX });
              this.handleEnd(e);
            }}
            onTouchStart={this.handleStart}
            onTouchEnd={this.handleEnd}
            trackRef={this.track}
            disabled={this.props.disabled!}
            selectionWidth={position - this.state.grab}
          />
          <Tooltip
            placement="bottom"
            overlay={this.state.value}
            trigger={['hover']}
          >
            <SlideHandler
              size={this.props.size!}
              innerRef={this.handle}
              style={{
                left: `${position}px`,
              }}
              onMouseDown={this.handleStart}
            />
          </Tooltip>
        </SliderWrapper>
        {this.renderIncreaseButton()}
      </SliderRoot>
    );
  }
}
