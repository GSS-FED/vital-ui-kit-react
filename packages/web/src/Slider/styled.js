import styled from 'styled-components';
import { Button as ButtonBase } from '../Button';
import constants from './constants';

const Root = styled.div`
  display: flex;
  position: relative;
  min-width: 200px;
  align-items: center;

  ${props =>
    props.disabled &&
    `
    opacity: .6;
  `};
`;

const Wrapper = styled.div`
  display: block;
  position: relative;
  width: 100%;
  height: ${props => constants[props.size].handlerSize};
`;

const Handler = styled.div`
  top: 0;
  left: ${props => `${props.left}px`}
  position: absolute;
  border-radius: 100%;
  border-color: #c4d2eb;
  background-color: #ffffff;
  box-shadow: 0 0 0 1px rgba(196, 210, 235, 0.3);
  text-indent: -3333px;
  overflow: hidden;
  text-decoration: none;
  text-align: center;
  outline: 0;
  border-style: solid;
  border-width: 1px;
  width: ${props =>
     constants[props.size].handlerSize};
  height: ${props => constants[props.size].handlerSize};
  cursor: pointer;

  &:hover {
    border-color: #C4D2EB;
    background-color: #FFFFFF;
    box-shadow: 0 0 0 2px rgba(196, 210, 235, 0.3);
  }
`;

const Button = styled(ButtonBase)`
  flex: 0 0 auto;
  width: 24px;
  height: 24px;
`;

export {
  Root,
  Wrapper,
  Handler,
  Button,
}
