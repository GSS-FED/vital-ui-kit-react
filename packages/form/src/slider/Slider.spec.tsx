import React from 'react';
import rerender from 'react-test-renderer';
import { Slider } from './Slider';

describe('Slider', () => {
  it('should render with default styles', () => {
    const slider = rerender.create(
      <Slider
        max={2}
        min={1}
        step={0.01}
        onChange={() => {}}
        value={1}
      />,
    );
    expect(slider.toJSON()).toMatchSnapshot();
  });
});
