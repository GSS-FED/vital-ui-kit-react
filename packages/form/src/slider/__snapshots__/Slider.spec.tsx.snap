// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider should render with default styles 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  min-width: 200px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: block;
  position: relative;
  width: 100%;
  height: 26px;
}

.c4 {
  top: 0;
  position: absolute;
  border-radius: 100%;
  border-color: #C3D1EB;
  background-color: #ffffff;
  box-shadow: 0 0 0 1px rgba(196,210,235,0.3);
  text-indent: -3333px;
  overflow: hidden;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  outline: 0;
  border-style: solid;
  border-width: 1px;
  width: 26px;
  height: 26px;
  cursor: pointer;
}

.c4:hover {
  border-color: #0E86FE;
  background-color: #ffffff;
  box-shadow: 0 0 0 2px rgba(196,210,235,0.3);
}

.c3 {
  position: absolute;
  margin: 0;
  padding: 0;
  border-radius: 5px;
  left: 0;
  height: 10px;
  top: 0;
  pointer-events: none;
  background-color: #0E86FE;
  border-bottom-right-radius: 0px;
  border-top-right-radius: 0px;
}

.c2 {
  position: relative;
  margin: 0;
  padding: 0;
  top: 50%;
  left: 0;
  height: 10px;
  margin-top: -4px;
  background-repeat: repeat-x;
  border-radius: 5px;
  border-color: #95ABD3;
  width: 100%;
  background-color: #D8E2F5;
  cursor: pointer;
}

<div
  aria-valuemax={2}
  aria-valuemin={1}
  aria-valuenow={1}
  className="c0"
  disabled={false}
>
  <div
    className="c1"
    disabled={false}
    size="medium"
  >
    <div
      className="c2"
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      size="medium"
    >
      <div
        className="c3"
        disabled={false}
        size="medium"
        style={
          Object {
            "width": 0,
          }
        }
        width={0}
      />
    </div>
    <div
      className="c4"
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      size="medium"
      style={
        Object {
          "left": "0px",
        }
      }
    />
    <div
      style={
        Object {
          "display": "none",
          "height": 0,
          "position": "absolute",
          "visibility": "hidden",
          "width": 0,
        }
      }
    />
  </div>
</div>
`;
